group/assignment5/Graph.java

```diff
@@ -0,0 +1,241 @@
+/**
+ * Assumption : the graph has vertices whose value is a character
+ */
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Stack;
+
+public class Graph {
+  private List<Vertex> vertices;
+
+  public Graph() {
+    vertices = new ArrayList<>();
+  }
+
+  /**
+   * Returns all vertices in the current graph.
+   *
+   * @return the list of the vertices in the current graph
+   */
+  public List<Vertex> getVertices() {
+    return vertices;
+  }
+
+  /**
+   * Adds a vertex to the graph if it does not already exist in the graph.
+   *
+   * @param vertex the vertex to be added.
+   * @return true, if the edge was successfully added, false if it already existed
+   * @throws IllegalArgumentException if the vertex to be added is null.
+   */
+  public boolean addVertex(Vertex vertex) {
+    if (vertex == null) {
+      throw new IllegalArgumentException();
+    }
+    if (!vertices.contains(vertex)) {
+      vertices.add(vertex);
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Returns the vertex that has the given character as value.
+   *
+   * @param value the value of interest
+   * @return the vertex that has the given value, or null if the vertex is inexistent
+   */
+  public Vertex getVertex(char value) {
+    for (Vertex v : vertices) {
+      if (v.getValue() == value) {
+        return v;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Topologically sorts the vertices using Kahn's non-recursive algorithm.
+   *
+   * @return one of the possible topological orders of the vertices.
+   */
+  public List<Character> topologicalSort() {
+    List<Character> order = new ArrayList<>();
+    LinkedList<Vertex> processNext = new LinkedList<>();
+    for (Vertex vertex : getVertices()) {
+      if (vertex.getInbound() == 0) {
+        processNext.add(vertex);
+      }
+    }
+
+    while (!processNext.isEmpty()) {
+      Vertex currentVertex = processNext.poll();
+      order.add(currentVertex.getValue());
+      for (Vertex neighbour : currentVertex.getNeighbours()) {
+        neighbour.decrementInbound();
+        if (neighbour.getInbound() == 0) {
+          processNext.add(neighbour);
+        }
+      }
+    }
+    
+    // no vertex inbound should be null at this point if the graph is acyclic
+    for (Vertex vertex : vertices) {
+      if (vertex.getInbound() != 0) {
+        return null;
+      }
+    }
+    return order;
+  }
+
+  /**
+   * Topologically sorts the vertices.
+   *
+   * @return all possible topological orders of the vertices.
+   */
+  public List<List<Character>> topologicalSorting() {
+    List<List<Character>> allOrders = new ArrayList<>();
```

Comment:
Maybe it's just me, but whenever I need only unique values, I prefer using sets.